### PR TITLE
libbpf-tools: convert wakeuptime tool

### PR DESCRIPTION
**Pull request by sterchelen**
Convert wakeuptime python tool to co-re version.

### CODE REVIEW

File: libbpf-tools/wakeuptime.bpf.c [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// Copyright (c) 2022 Nicolas Sterchele
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "wakeuptime.h"
+#include "maps.bpf.h"
+
+#define PF_KTHREAD		0x00200000	/* kernel thread */
+#define MAX_ENTRIES		10240
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.bpf.c:11]: This is different from header file ?

File: libbpf-tools/wakeuptime.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+
+#ifndef __WAKEUPTIME_H
+#define __WAKEUPTIME_H
+
+#define MAX_ENTRIES 102400
+#define TASK_COMM_LEN 16
+
+struct key_T {
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.h:9]: This naming is weird, just `key` or `key_t` ?

File: libbpf-tools/wakeuptime.bpf.c [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    if (tsp == 0)
+        return 0;        // missed start
+    bpf_map_delete_elem(&start, &tid);
+
+    delta = bpf_ktime_get_ns() - *tsp;
+    delta = delta / 1000;
+    if ((delta < min_block_ns) || (delta > max_block_ns))
+        return 0;
+
+    struct key_T key = {};
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.bpf.c:74]: We are following kernel code style, so declare this variable on the top of the function.

File: libbpf-tools/wakeuptime.bpf.c (hunk 1 above)
Comment (sterchelen) [reply libbpf-tools/wakeuptime.bpf.c]: fixed in 5209cddb8a68d203a0b95eb8622150e5068c0265

File: libbpf-tools/wakeuptime.h (hunk 2 above)
Comment (sterchelen) [reply libbpf-tools/wakeuptime.h]: fixed in 5209cddb8a68d203a0b95eb8622150e5068c0265

File: libbpf-tools/wakeuptime.bpf.c [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(key_size, sizeof(u32));
+} stackmap SEC(".maps");
+
+static __always_inline
+int offcpu_sched_switch(struct task_struct *prev) {
+    u64 pid_tgid = bpf_get_current_pid_tgid();
```
[the remark this answers is not in the document]
Comment (sterchelen) [reply libbpf-tools/wakeuptime.bpf.c]: fixed in 5209cddb8a68d203a0b95eb8622150e5068c0265

File: libbpf-tools/wakeuptime.bpf.c (hunk 3 above)
Comment (sterchelen) [reply libbpf-tools/wakeuptime.bpf.c]: fixed in 5209cddb8a68d203a0b95eb8622150e5068c0265

**Comment by sterchelen** [comment]
The re-request review button doesn't work... then I ask a re-request review by message @chenhengqi, please.

File: libbpf-tools/wakeuptime.bpf.c [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+} start SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(key_size, sizeof(u32));
+} stackmap SEC(".maps");
+
+static __always_inline
+int offcpu_sched_switch(struct task_struct *prev) {
+		u64 pid_tgid = bpf_get_current_pid_tgid();
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.bpf.c:38]: * You don't need `__always_inline`
* `{` on a single line
* Use one tab for indention

File: libbpf-tools/wakeuptime.bpf.c [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+				return 0;
+
+		key.w_k_stack_id = bpf_get_stackid(ctx, &stackmap, 0);
+		bpf_probe_read_kernel(&key.target, sizeof(key.target), p->comm);
+		bpf_get_current_comm(&key.waker, sizeof(key.waker));
+
+		count_key = bpf_map_lookup_or_try_init(&counts, &key, &zero);
+		if (count_key) {
+			__atomic_add_fetch(count_key, delta, __ATOMIC_RELAXED);
+		}
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.bpf.c:81]: Remove `{}` here.

File: libbpf-tools/wakeuptime.c [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+static void print_map(struct ksyms *ksyms, struct wakeuptime_bpf *obj)
+{
+	struct key_t lookup_key = {}, next_key;
+	int err, i, counts_fd, stack_traces_fd, val;
+	unsigned long *ip;
+	const struct ksym *ksym;
+
+	ip = calloc(env.perf_max_stack_depth, sizeof(*ip));
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.c:166]: Missing `free(ip)`.

Comment (sterchelen) [reply libbpf-tools/wakeuptime.c:166]: fixed in 8efb198

File: libbpf-tools/wakeuptime.bpf.c (hunk 6 above)
Comment (sterchelen) [reply libbpf-tools/wakeuptime.bpf.c]: fixed in 8efb198

File: libbpf-tools/wakeuptime.bpf.c (hunk 5 above)
Comment (sterchelen) [reply libbpf-tools/wakeuptime.bpf.c]: fixed in 8efb198

File: libbpf-tools/wakeuptime.bpf.c [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+	if (targ_pid && targ_pid != pid)
+		return 0;
+	tsp = bpf_map_lookup_elem(&start, &tid);
+	if (tsp == 0)
+		return 0;	// missed start
+	bpf_map_delete_elem(&start, &tid);
+
+	delta = bpf_ktime_get_ns() - *tsp;
+	delta = delta / 1000;
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.bpf.c:72]: Do the division in userspace, save cycles in BPF side.

File: libbpf-tools/wakeuptime.c [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+	libbpf_set_print(libbpf_print_fn);
+
+	obj = wakeuptime_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open BPF object\n");
+		return 1;
+	}
+
+	obj->rodata->targ_pid = env.pid;
+	obj->rodata->min_block_ns = env.min_block_time;
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.c:234]: `env.min_block_time * 1000` ?

**Comment by sterchelen** [comment]
@chenhengqi, new version with fixes requested here 61b39bdda3c14868739fac0d759014ec7a119cb3

File: libbpf-tools/wakeuptime.c (hunk 9 above)
Comment (sterchelen) [reply libbpf-tools/wakeuptime.c]: Indeed, offcputime seems also wrong... didn't see a conversion statement 👇🏼 
https://github.com/iovisor/bcc/blob/f6c4b4c618a1a471fdfdd8bc852578f383793b63/libbpf-tools/offcputime.c#L297-L298

**Comment by chenhengqi** [comment]
You should remove `delta = delta / 1000;` from BPF program.

Other than that, LGTM.

**Comment by sterchelen** [comment]
> You should remove delta = delta / 1000; from BPF program.

Fixed @chenhengqi 👍🏼

**Review by chenhengqi — APPROVED** [review]
LGTM.

**Comment by sterchelen** [comment]
@chenhengqi Do you know who is responsible to merge libbpf-tools PR?

**Comment by chenhengqi** [comment]
cc @davemarchevsky

**Comment by sterchelen** [comment]
Any news on this @davemarchevsky?

File: libbpf-tools/wakeuptime.c [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+		err = bpf_map_lookup_elem(stack_traces_fd, &next_key.w_k_stack_id, ip);
+		if (err < 0) {
+			fprintf(stderr, "missed kernel stack: %d\n", err);
+		}
+		for (i = 0; i < env.perf_max_stack_depth && ip[i]; i++) {
+			ksym = ksyms__map_addr(ksyms, ip[i]);
+			printf("	%-16lx %s\n", ip[i], ksym ? ksym->name: "Unknown");
+		}
+		printf("	%16s %s\n","waker:", next_key.waker);
+		//to convert val in microseconds
```
Comment (chenhengqi) [inline libbpf-tools/wakeuptime.c:194]: Nit: No C++ style comment.

**Comment by chenhengqi** [comment]
@sterchelen Please rebase and push again.

**Comment by sterchelen** [comment]
> @sterchelen Please rebase and push again.

Done @chenhengqi

**Pull request merged**